aws-lc-rs/src/cipher.rs

```diff
@@ -3,158 +3,1268 @@
+//! # use std::error::Error;
+//! #
+//! # fn main() -> Result<(), Box<dyn Error>> {
+//! use aws_lc_rs::cipher::{DecryptingKey, EncryptingKey, UnboundCipherKey, AES_128};
+//!
+//! let original_message = "This is a secret message!".as_bytes();
+//! let mut in_out_buffer = Vec::from(original_message);
+//!
+//! let key_bytes: &[u8] = &[
+//!     0xff, 0x0b, 0xe5, 0x84, 0x64, 0x0b, 0x00, 0xc8, 0x90, 0x7a, 0x4b, 0xbf, 0x82, 0x7c, 0xb6,
+//!     0xd1,
+//! ];
+//!
+//! let key = UnboundCipherKey::new(&AES_128, key_bytes)?;
+//! let encrypting_key = EncryptingKey::ctr(key)?;
+//! let context = encrypting_key.encrypt(&mut in_out_buffer)?;
+//!
+//! let key = UnboundCipherKey::new(&AES_128, key_bytes)?;
+//! let decrypting_key = DecryptingKey::ctr(key, context)?;
+//! let plaintext = decrypting_key.decrypt(&mut in_out_buffer)?;
+//! assert_eq!(original_message, plaintext);
+//! #
+//! # Ok(())
+//! # }
+//! ```
+//!
+
+#![allow(clippy::module_name_repetitions)]
+
 pub(crate) mod aes;
 pub(crate) mod block;
 pub(crate) mod chacha;
+pub(crate) mod key;
 
-use crate::cipher::aes::{encrypt_block_aes, Aes128Key, Aes256Key};
-use crate::cipher::block::Block;
-use crate::cipher::chacha::ChaCha20Key;
 use crate::error::Unspecified;
-use aws_lc::{AES_set_encrypt_key, AES_KEY};
-use std::mem::{size_of, transmute, MaybeUninit};
-use std::os::raw::c_uint;
-use std::ptr;
+use crate::hkdf;
+use crate::hkdf::KeyType;
+use crate::iv::FixedLength;
+use aws_lc::{AES_cbc_encrypt, AES_ctr128_encrypt, AES_DECRYPT, AES_ENCRYPT, AES_KEY};
+use key::SymmetricCipherKey;
+use std::fmt::Debug;
+use std::mem::MaybeUninit;
 use zeroize::Zeroize;
 
-pub(crate) enum SymmetricCipherKey {
-    Aes128(Aes128Key, AES_KEY),
-    Aes256(Aes256Key, AES_KEY),
-    ChaCha20(ChaCha20Key),
+/// The cipher block padding strategy.
+#[non_exhaustive]
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum PaddingStrategy {
+    /// PKCS#7 Padding. ([See RFC 5652](https://datatracker.ietf.org/doc/html/rfc5652#section-6.3))
+    PKCS7,
+}
+
+impl PaddingStrategy {
+    fn add_padding<InOut>(self, block_len: usize, in_out: &mut InOut) -> Result<(), Unspecified>
+    where
+        InOut: AsMut<[u8]> + for<'in_out> Extend<&'in_out u8>,
+    {
+        match self {
+            PaddingStrategy::PKCS7 => {
+                let in_out_len = in_out.as_mut().len();
+                // This implements PKCS#7 padding scheme, used by aws-lc if we were using EVP_CIPHER API's
+                let remainder = in_out_len % block_len;
+                if remainder == 0 {
+                    let block_size: u8 = block_len.try_into().map_err(|_| Unspecified)?;
+                    in_out.extend(vec![block_size; block_len].iter());
+                } else {
+                    let padding_size = block_len - remainder;
+                    let v: u8 = padding_size.try_into().map_err(|_| Unspecified)?;
+                    // Heap allocation :(
+                    in_out.extend(vec![v; padding_size].iter());
```

Comment:
It would apply to both branches. 

I updated the implementation in the latest commit to use a 16-byte `padding_buffer` instead of the `vec!` macro. With this change it's possibly not a "heap allocation" depending on the "capacity" of the collection and how it implements the [`Extend`](https://dev-doc.rust-lang.org/beta/std/iter/trait.Extend.html) trait.